src/puppetlabs/dujour/version_check.clj

```diff
@@ -0,0 +1,125 @@
+  "Make a request to the puppetlabs server to determine the latest available
+  version. Returns the JSON object received from the server, which
+  is expected to be a map containing keys `:version`, `:newer`, and `:link`.
+  Returns `nil` if the request does not succeed for some reason."
+  [request-values :- RequestValues
+   update-server-url :- schema/Str]
+  (let [product-name (:product-name request-values)
+        {:keys [group-id artifact-id]} (get-coords product-name)
+        current-version (version group-id artifact-id)
+        version-data {:version current-version}
+        query-string (ring-codec/form-encode version-data)
+        url (format "%s?product=%s&group=%s&%s" update-server-url artifact-id group-id query-string)
+        {:keys [status body] :as resp} (client/get url
+                                                   {:headers {"Accept" "application/json"}
+                                                    :as :text})]
+    (if (= status 200)
+      (json/parse-string body true)
+      (sling/throw+ {:type ::update-request-failed
+                     :message resp}))))
+
+
+(defn validate-config!
+  [request-values update-server-url]
+  ;; if this ends up surfacing error messages that aren't very user-friendly,
+  ;; we can improve the validation logic.
+  (schema/validate RequestValues request-values)
+  (schema/validate (schema/maybe schema/Str) update-server-url))
+
+(defn version-check
+  "This will fetch the latest version number and log if the system
+  is out of date."
+  [request-values update-server-url]
+  (log/debugf "Checking for newer versions of %s" (:product-name request-values))
+  (let [update-server-url             (or update-server-url default-update-server-url)
+        {:keys [version newer link]}  (try
+                                        (update-info request-values update-server-url)
+                                        (catch Throwable e
+                                          (log/debug e (format "Could not retrieve update information (%s)" update-server-url))))
+        link-str (if link
+                   (format " Visit %s for details." link)
+                   "")
+        update-msg (format "Newer version %s is available!%s" version link-str)]
+    (when newer
+      (log/info update-msg))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Public
+
+(defn check-for-updates!
+  ([request-values update-server-url]
+    (check-for-updates! request-values update-server-url nil))
+  ([request-values update-server-url callback-fn]
+    (validate-config! request-values update-server-url)
+    (future
+      (try
+        (version-check request-values update-server-url)
+        (catch Exception e
+          (log/warn e "Error occurred while checking for updates")
+          (throw e))
+        (finally
+          (if-not (nil? callback-fn)
+            (callback-fn)))))
```

Comment:
I think it might be nice for the callback-fn's to take a single map as an argument.  So you could have the `version-check` function return the map that it got back in the server response, and then pass that through as an argument to the callback function?